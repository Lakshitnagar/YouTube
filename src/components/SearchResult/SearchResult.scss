.search{
  padding: 16px;
}

.search__header{
  padding: 16px;
}

.search__result {
  display: flex;
  background-color: #f9f9f9;
  padding: 16px;
  flex-wrap: wrap;
  align-content: flex-start;

  &::after {
    content: '';
    flex-grow: 1000000000;
  }
}

.searchError {
  margin: auto;
  padding: 0 16px;
  font-size: xx-large;
  color: darkgoldenrod;
}
