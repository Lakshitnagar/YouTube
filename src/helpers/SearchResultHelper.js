export const getVideoList = (searchResult)=>{
    return searchResult ? searchResult.items : [];
};

export const getNextPageToken = (searchResult)=>{
    return searchResult ? searchResult.nextPageToken : '';
};
