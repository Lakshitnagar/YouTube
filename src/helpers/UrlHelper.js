const searchUrl = "https://www.googleapis.com/youtube/v3/search";

export const getYTSearchUrl = () => {
    return searchUrl;
};
