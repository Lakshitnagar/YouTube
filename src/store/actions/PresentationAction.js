import {createAction} from "@reduxjs/toolkit";
import {ADD_API_FAILURE_STATUS, REMOVE_API_FAILURE_STATUS} from "../StoreConstants";

export const addApiFailureStatus = createAction(ADD_API_FAILURE_STATUS);
export const removeApiFailureStatus = createAction(REMOVE_API_FAILURE_STATUS);
