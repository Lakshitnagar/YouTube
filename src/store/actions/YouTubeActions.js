import {createAction} from "@reduxjs/toolkit";
import {
    RESET_YOU_TUBE_SEARCH_KEYWORD,
    RESET_YOU_TUBE_SEARCH_RESULTS,
    SET_YOU_TUBE_SEARCH_KEYWORD,
    SET_YOU_TUBE_SEARCH_RESULTS
} from "../StoreConstants";
import {searchByKeyword} from "../../http/YoutubeClient";
import {addApiFailureStatus, removeApiFailureStatus} from "./PresentationAction";
import {YOUTUBE_SEACRH_API} from "../../constants/ApiConstants";

export const setYouTubeSearchResults = createAction(SET_YOU_TUBE_SEARCH_RESULTS);
export const resetYouTubeSearchResults = createAction(RESET_YOU_TUBE_SEARCH_RESULTS);

export const setYouTubeSearchKeyword = createAction(SET_YOU_TUBE_SEARCH_KEYWORD);
export const resetYouTubeSearchKeyword = createAction(RESET_YOU_TUBE_SEARCH_KEYWORD);

export const fetchYouTubeVideosByKeyword = (keyword) => {
    return async (dispatch) => {
        let videosResponse = null;

        dispatch(setYouTubeSearchKeyword(keyword));
        dispatch(removeApiFailureStatus(YOUTUBE_SEACRH_API));
        try {
            videosResponse = await searchByKeyword(keyword);
            dispatch(setYouTubeSearchResults(videosResponse));
        } catch (error) {
            dispatch(addApiFailureStatus(YOUTUBE_SEACRH_API));
        } finally {
        }
    };
};
