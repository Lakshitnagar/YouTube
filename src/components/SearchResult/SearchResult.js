import React from 'react';
import './SearchResult.scss';
import VideoListItem from "../VideoListItem/VideoListItem";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {getNextPageToken, getVideoList} from "../../helpers/SearchResultHelper";
import {YOUTUBE_SEACRH_API} from "../../constants/ApiConstants";
import {fetchYouTubeVideosByKeywordNextPage} from "../../store/actions/YouTubeActions";

export function SearchResult(props) {
    const searchOnScrollHandler = (e) => {
        if (e.target.scrollTop + e.target.clientHeight >= e.target.scrollHeight) {
            props.fetchYouTubeVideosByKeywordNextPage(props.searchKeyword, props.nextPageToken);
        }
    };

    if (props.isError) {
        return (
            <div className="searchError">
                Ahh.. Something is not right !!
            </div>
        );
    }

    return (
        <div className="search" onScroll={searchOnScrollHandler}>
            {props.searchKeyword && <div className="search__header">Showing results for "{props.searchKeyword}"</div>}
            <div className="search__result">
                {props.videoList.length !== 0 && props.videoList.map((videoDetail, ind) => {
                    return <VideoListItem key={ind} videoDetail={videoDetail}/>
                })}
                {!props.videoList.length && <div>Search videos by any keyword</div>}
            </div>
        </div>
    );
}

export const mapStateToProps = (state) => {
    return {
        videoList: getVideoList(state.youtubeSearchResults),
        nextPageToken: getNextPageToken(state.youtubeSearchResults),
        searchKeyword: state.youtubeSearchKeyword,
        isError: state.presentationConfig.apiStatus.includes(YOUTUBE_SEACRH_API)
    };
};

export const mapDispatchToProps = (dispatch) => {
    return {
        fetchYouTubeVideosByKeywordNextPage: (keyword, nextPageToken) => dispatch(fetchYouTubeVideosByKeywordNextPage(keyword, nextPageToken))
    };
};

SearchResult.propTypes = {
    videoList: PropTypes.array,
    searchKeyword: PropTypes.string,
    nextPageToken: PropTypes.string,
    isError: PropTypes.bool,
    fetchYouTubeVideosByKeywordNextPage: PropTypes.func
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchResult);
