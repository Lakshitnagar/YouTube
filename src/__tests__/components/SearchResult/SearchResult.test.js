import React from 'react';
import {shallow} from "enzyme";
import {shallowToJson} from "enzyme-to-json";
import {mapStateToProps, SearchResult} from "../../../components/SearchResult/SearchResult";
import * as SearchResultHelper from "../../../helpers/SearchResultHelper";
import * as YouTubeActions from "../../../store/actions/YouTubeActions";
import {mapDispatchToProps} from "../../../components/SearchResult/SearchResult";

const mockFetchYouTubeVideosByKeywordNextPage = jest.fn();

describe('SearchResult', function () {

    it('should render and match snapshot', function () {
        const component = <SearchResult videoList={[{}, {}, {}]} searchKeyword={'hello kitty'}/>;

        const wrapper = shallow(component);

        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should render and match snapshot on error', function () {
        const component = <SearchResult isError={true}/>;

        const wrapper = shallow(component);

        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should render and match snapshot on empty result', function () {
        const component = <SearchResult videoList={[]}/>;

        const wrapper = shallow(component);

        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should not fetch next page results if not scrolled to bottom', function () {
        const component = <SearchResult videoList={[]} fetchYouTubeVideosByKeywordNextPage={mockFetchYouTubeVideosByKeywordNextPage}/>;

        const wrapper = shallow(component);

        wrapper.find('.search').simulate('scroll', {
            target: {
                scrollTop: 25,
                clientHeight: 50,
                scrollHeight: 100
            }
        });

        expect(mockFetchYouTubeVideosByKeywordNextPage).toHaveBeenCalledTimes(0);
    });

    it('should fetch next page results scrolled to bottom', function () {
        const component = <SearchResult videoList={[]} fetchYouTubeVideosByKeywordNextPage={mockFetchYouTubeVideosByKeywordNextPage}
        searchKeyword={'keyword'} nextPageToken={'nextPageToken'}/>;

        const wrapper = shallow(component);

        wrapper.find('.search').simulate('scroll', {
            target: {
                scrollTop: 50,
                clientHeight: 50,
                scrollHeight: 100
            }
        });

        expect(mockFetchYouTubeVideosByKeywordNextPage).toHaveBeenCalledWith('keyword', 'nextPageToken');
    });
});

describe('mapStateToProps', function () {
    beforeEach(()=>{
        spyOn(SearchResultHelper, 'getVideoList').and.returnValue('VIDEO_LIST');
        spyOn(SearchResultHelper, 'getNextPageToken').and.returnValue('NEXT_PAGE_TOKEN');
    });

    it('should add some processed state variables to props', function () {
        const state = {
            youtubeSearchResults: {},
            youtubeSearchKeyword: 'SEARCH_KEYWORD',
            presentationConfig: {
                apiStatus: ['YOUTUBE_SEACRH_API']
            }
        };
        const actual = mapStateToProps(state);

        expect(actual).toEqual({
            videoList: 'VIDEO_LIST',
            nextPageToken: 'NEXT_PAGE_TOKEN',
            searchKeyword: 'SEARCH_KEYWORD',
            isError: true
        });
    });
});

describe('mapDispatchToProps', () => {
    it('should add dispatches to props', () => {
        const dispatch = jest.fn();
        spyOn(YouTubeActions, 'fetchYouTubeVideosByKeywordNextPage').and.returnValue(mockFetchYouTubeVideosByKeywordNextPage);

        const actual = mapDispatchToProps(dispatch);

        actual.fetchYouTubeVideosByKeywordNextPage('hello', 'nextPageToken');

        expect(Object.getOwnPropertyNames(actual)).toEqual(['fetchYouTubeVideosByKeywordNextPage']);
        expect(dispatch).toHaveBeenCalledWith(mockFetchYouTubeVideosByKeywordNextPage);
    });
});
