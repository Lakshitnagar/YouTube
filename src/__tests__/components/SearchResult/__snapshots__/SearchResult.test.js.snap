// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResult should render and match snapshot 1`] = `
<div
  className="search"
  onScroll={[Function]}
>
  <div
    className="search__header"
  >
    Showing results for "
    hello kitty
    "
  </div>
  <div
    className="search__result"
  >
    <VideoListItem
      key="0"
      videoDetail={Object {}}
    />
    <VideoListItem
      key="1"
      videoDetail={Object {}}
    />
    <VideoListItem
      key="2"
      videoDetail={Object {}}
    />
  </div>
</div>
`;

exports[`SearchResult should render and match snapshot on empty result 1`] = `
<div
  className="search"
  onScroll={[Function]}
>
  <div
    className="search__result"
  >
    <div>
      Search videos by any keyword
    </div>
  </div>
</div>
`;

exports[`SearchResult should render and match snapshot on error 1`] = `
<div
  className="searchError"
>
  Ahh.. Something is not right !!
</div>
`;
